cep-abi3.md

```diff
@@ -0,0 +1,269 @@
+
+An example for a `noarch: python` recipe.
+
+```json
+{
+  "arch": null,
+  "build": "pyh2585a3b_103",
+  "build_number": 103,
+  "depends": [
+    "mpmath >=0.19",
+    "python >=3.8"
+  ],
+  "license": "BSD-3-Clause",
+  "license_family": "BSD",
+  "name": "sympy",
+  "noarch": "python",
+  "platform": null,
+  "subdir": "noarch",
+  "timestamp": 1718625708903,
+  "version": "1.12.1"
+}
+```
+
+### Current behaviour in solver tools
+
+Conda package upload tools like `anaconda-client` use `A1` to upload
+the package to the `noarch` subdir.
+
+Conda install tools have slightly different behaviour.
+
+Conda:
+1. Actions `B1, B2, B3` are applied for packages with `A3`.
+2. Action `B4` is applied for packages with `A4`.
+
+Micromamba:
+1. Actions `B1, B2, B3` are applied for packages with both `A2, A3`.
+2. Action `B4` is applied for packages with both `A2, A4`.
+
+
+## Implementation for abi3 packages in install tools.
+
+In order to support abi3 packages, we propose two methods.
+
+### `package_metadata_version = 1`
+
+We require the following attributes in abi3 packages:
+
+&nbsp;&nbsp;<strong>C1</strong>:
+  They have `subdir: <platform>` where `<platform>` is the subdir
+  that the package was built for.
+
+&nbsp;&nbsp;<strong>C2</strong>:
+  They have `noarch: python`.
+
+&nbsp;&nbsp;<strong>C3</strong>:
+  `A2, A3, A4` are applied.
+
+This is compatible with `conda/mamba/micromamba` install tools
+currently. This requires support from build tools to set `subdir: <platform>`
+only.
+
+In particular an option:
+```
+build:
+  python_version_independent: true
+```
+to imply a python version independent package and set `noarch: python`
+in `info/index.json`.
+
+To test that an install tool works with this scheme, you can try
+
+    conda/micromamba create -n a sympy python=3.10 -c isuruf/label/abi3 -c conda-forge
+
+on a linux-64 machine and confirm that all actions `B1, B2, B3, B4` are applied.
+
+### `package_metadata_version = 2`
+
+In this method, we require additional support from install tools.
+
+For `abi3` and `noarch: python` packages, we record the `entry_points` and
+the pure python files in `info/link.json`.
+
+```json
+{
+  "python": {
+    "entry_points": [
+      "foo = foo:main"
+    ],
+    "py_compile": [
+      "lib/python/site-packages/foo/main.py",
+      "lib/python/site-packages/foo/__init__.py",
+    ]
+  },
+  "package_metadata_version": 2,
+}
+```
+
+We require the following support from install tools
+
+&nbsp;&nbsp;<strong>D1</strong>:
+  Apply action `B4` if `python: entry_points` is present in `info/link.json`.
+
+&nbsp;&nbsp;<strong>D2</strong>:
+  Apply actions `B2, B3` if `python: py_compile` is present in `info/link.json`.
+
+&nbsp;&nbsp;<strong>D3</strong>:
+  Provide a `__supports_package_metadata_version=2` virtual package.
+
+Note that we do not require `B1` as package authors should depend on a python
+version that has a custom `site.py` that adds `lib/python/site-packages` to
+the path.
```

Comment:
Doesnt this imply that older python versions available on for instance conda-forge will never have the ability to use abi3 packages? If we are modifying the installer we could just as well move the Python files to the python specific site-packages directory.

Reply:
Sure, but we can also have the `site.py` in a separate conda package as well. I'm trying to reduce all these special things that a installer has to do. With the files in `lib/python/site-packages`, the only thing that a installer has to do for an abi3 package is to compile the pyc files which is an optional thing anyway.

Reply:
I'm hoping we can add support in defaults too.

Reply:
I would not prefer this because it means the proper functioning of the installer is dependent on a specific package. I'd rather encode the behavior in the installer itself.

Reply:
`lib/python3.10/site-packages` is a CPython specific directory and we had to use that for PyPy too because the installer didn't know about PyPy.

Can you explain why you want the installer to do special things? I would think that it's better for the installer to do less special case stuff.

Reply:
I agree that the installer should do less special case stuff. But I also don't think that it should depend on specific packages to function properly.

Even with this proposal the installer does already have special case behavior. However, without the dependency on a package that introduces a `site.py` file it will leave the package in a broken state. To me that adds complexity for the package maintainer.

However, one could argue that the same holds for depending on `python` itself.. 

It would be ideal if a recipe author doesn't have to care about these details.

Reply:
We could make the build tool add this package, then the installer does not have to care about it.

Reply:
Do we do this for other packages? Because then we tie the build-tool to specific package names which might also not be ideal.

Reply:
We take the C compiler name from `c_compiler` variable in the config. We could do the same.

Reply:
Oh I forgot about `run_exports: noarch`. We can extend that in the python recipe.